File: algorithms/two-sum/javascript/two-sum.js
// solution 1
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (nums, target) {
  for (let i = 0; i < nums.length; ++i) {
    for (let j = i + 1; j < nums.length; ++j) {
      if (nums[i] + nums[j] === target) {
        return [i, j];
      }
    }
  }
};

// solution 2
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = function (nums, target) {
  let map = new Map();
  nums.forEach((value, index) => {
    map.set(value, index);
  });

  for (let i = 0; i < nums.length; i++) {
    let complement = target - nums[i];

    if (map.has(complement) && i !== map.get(complement)) {
      return [map.get(complement), i];
    }
  }
};

File: algorithms/two-sum/two-sum.cpp
class Solution
{
public:
    // solution 1
    vector<int> twoSum(vector<int> &nums, int target)
    {
        for (int i = 0; i < nums.size(); i++)
        {
            for (int j = i + 1; j < nums.size(); j++)
            {
                if (nums[i] + nums[j] == target)
                {
                    return {i, j};
                }
            }
        }
        return {};
    }

    // solution 2
    vector<int> twoSum(vector<int> &nums, int target)
    {
        unordered_map<int, int> map = {};

        for (int i = 0; i < nums.size(); ++i)
        {
            map[nums[i]] = i;
        }

        for (int i = 0; i < nums.size(); i++)
        {
            int complement = target - nums[i];
            if (map.find(complement) != map.end() && map[complement] != i)
            {
                return {map[complement], i};
            }
        }
        return {};
    }
};